### PR TITLE
Metrics/flexibility

### LINKED ISSUE #4461: Internal: Filter metrics by namespaces
As a consumer of the metrics collector I aim to get meaningful namespaces that let me select and fetch data easy. For now we've:

* root.base.events_in
* root.base.events_filtered

I would propose to have something that matches more close, semantically speaking the resources defined for #4446, so we can make something like.

* root.stats.events, to get all events information
* root.stats.jvm, to get all jvm information
* root.stats.networking, to get all events information.

I also propose to rename root to a word with a bit more context, what do you think? might be something like metrics? or logstash? then we could have namespaces like:

* logstash.stats.events
* logstash.stats.jvm 

what do you think?

### LINKED ISSUE #4529: Hide the implementations details of the MetricStore when retrieving metrics
Hi,
  as a user of the metric store I don't expect to interact with implementation details, today when asking the metric store for data it returns an instance of a ```Concurrent::Map```, this makes the consumer of this api aware of an implementation detail that might change in the feature.

We should return a plain ruby hash, or similar.

The way the Metric store is implemented also make the values of this ```Concurrent::Map``` be a MetricType instance. This is problematic as to retrieve a value out of it a user must call it's value method. The returned value object for the metric store should be a plain ruby object or similar.

### PR DESCRIPTION
**Pull request by ph**
Refactoring of the metric:

- Conform to the metric defined in Follow #4492
- Better filtering of the metrics from the metric store Fixes #4461
- Hide the implementation details of the metric store It return an hash, Fixes #4529
- MetricStore#each now support a path to only fetch a subset on the metrics.

### CODE REVIEW

**Comment by ph** [comment]
:bug: Fixing the build. weird I did not have this before pushing.

File: logstash-core/lib/logstash/agent.rb [hunk 1]
```diff
@@ -22,7 +22,7 @@ class LogStash::Agent
 
   def initialize(options = {})
     @pipelines = {}
-     
+
```
Comment (purbon) [inline logstash-core/lib/logstash/agent.rb]: always found this changes very relevant, no just kidding sometimes git thinks we deleted and added a new line :panda_face:

File: logstash-core/lib/logstash/agent.rb [hunk 2]
```diff
@@ -106,7 +106,7 @@ def stop_background_services
   def configure_metric
     if collect_metric?
       @logger.debug("Agent: Configuring metric collection")
-      @metric = LogStash::Instrument::Metric.create(:root)
```
Comment (purbon) [inline logstash-core/lib/logstash/agent.rb]: Curious about it, As I understand we spoke about having as namespaces something like ```stats.events.in```, how does it much here the removal of ```root```? might be I lost it, but I found not references to stats anywhere :smiley_cat:

File: logstash-core/lib/logstash/filter_delegator.rb [hunk 3]
```diff
@@ -31,18 +34,26 @@ def config_name
...
     end
 
     private
     def define_flush_method
       define_singleton_method(:flush) do |options = {}|
-        @filter.flush(options)
+        # we also need to trace the number of events
+        # coming from a specific filters.
+        new_events = @filter.flush(options)
+        @metric_events.increment(:out, new_events.size) unless new_events.nil? # Logstash-filter-aggregates return nil.
```
Comment (purbon) [inline logstash-core/lib/logstash/filter_delegator.rb]: re:
> # Logstash-filter-aggregates return nil

is this a common option ? could happen with other filters? what do you think?

File: logstash-core/lib/logstash/inputs/metrics.rb [hunk 4]
```diff
@@ -11,6 +11,7 @@ module LogStash module Inputs
   # Elasticsearch
   class Metrics < LogStash::Inputs::Base
     config_name "metrics"
+    milestone 3
```
Comment (purbon) [inline logstash-core/lib/logstash/inputs/metrics.rb]: I thought we were not using milestones anymore :+1:

File: logstash-core/lib/logstash/inputs/metrics.rb [hunk 5]
```diff
@@ -35,19 +36,22 @@ def stop
...
-      #   - We can use a synchronization mechanism between the called thread (update method)
-      #   and the plugin thread (run method)
-      #   - How we handle back pressure here?
-      #   - one snashot should be only one event ?
+      # The back pressure is handled in the collector's
+      # scheduled task (running into his own thread) if something append to one of the listener it will
+      # will timeout. In a sane pipeline, with a low traffic of events it shouldn't be a problems.
       snapshot.metric_store.each do |metric|
-        @queue << LogStash::Event.new({ "@timestamp" => snapshot.created_at }.merge(metric.to_hash))
+        @queue << LogStash::Event.new({ "@timestamp" => snapshot.created_at }.merge(convert_to_hash(metric)))
```
Comment (purbon) [inline logstash-core/lib/logstash/inputs/metrics.rb]: I wonder about this conversion happening here? would not be more coherent that the to_hash method gets implemented in the metric level? I guess it collide with the other sanitization, but the format specified in the ```convert_to_hash``` is not really doing rocket changes. What do you think?

File: logstash-core/lib/logstash/instrument/collector.rb [hunk 6]
```diff
@@ -19,7 +19,7 @@ class Collector
     include Observable
     include Singleton
 
-    SNAPSHOT_ROTATION_TIME_SECS = 10 # seconds
+    SNAPSHOT_ROTATION_TIME_SECS = 1 # seconds
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/collector.rb]: I know, this is still work in progress, but why this change again? I mean we changed from 1 to 10 and not back to 1, what is your thinking here?

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 7]
```diff
@@ -8,87 +8,102 @@ class MetricNoKeyProvided < MetricException; end
...
+    def initialize(collector)
       @collector = collector
-      @namespace_information = Array(namespace).map(&:to_sym)
     end
 
-    def increment(key, value = 1)
+    def increment(namespace, key, value = 1)
       validate_key!(key)
-      collector.push(namespace_information, key, :counter, :increment, value)
+      collector.push(namespace, key, :counter, :increment, value)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: not sure I like this change, does this means now we could use "any" metric object to run operations over other keys? I understand this provide us more flexibility, but also introduce more degrees of complexity, if we go that path, do we want to make sure the namespace is consistent before issuing the push?

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 8]
```diff
@@ -8,87 +8,102 @@ class MetricNoKeyProvided < MetricException; end
...
+    #   namespaced_metric.increment(:mykey, 200)
+    #   namespaced_metric.increment(:mykey_2, 200)
+    # ```
+    #
+    # @param name [Array<String>] Name of the namespace
+    # @param name [String] Name of the namespace
+    def namespace(name)
+      raise MetricNoNamespaceProvided if name.nil? || name.empty?
+
+      NamespacedMetric.new(self, name)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: oh! now I understand changes inside the Metric class, :smile:

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 9]
```diff
@@ -8,87 +8,102 @@ class MetricNoKeyProvided < MetricException; end
...
     end
 
-    def self.create(namespace, collector = LogStash::Instrument::Collector.instance)
-      Metric.new(collector, namespace)
+    # Create a Metric instrance using the default Collector singleton reference
+    #
+    #
+    def self.create(collector = LogStash::Instrument::Collector.instance)
+      Metric.new(collector)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: should all be ```NamespacedMetric```? what is the benefit of having the first one be a base Metric?

File: logstash-core/lib/logstash/instrument/metric.rb [hunk 10]
```diff
@@ -8,87 +8,102 @@ class MetricNoKeyProvided < MetricException; end
...
       raise MetricNoKeyProvided if key.nil? || key.empty?
     end
+
+    # Allow to calculate the execution of a block of code.
+    # This class support 2 differents syntax a block or the return of
+    # the object itself, but in the later case the metric wont be recorded
+    # Until we call `#stop`.
+    #
+    # @see LogStash::Instrument::Metric#time
+    class TimedExecution
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: I understand this class is under private scope, right? I wonder what happen in the case of the time function returning an instance of ```TimedExecution``` would this lead to some problem?

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 11]
```diff
@@ -8,7 +8,14 @@ module LogStash module Instrument
   # saved in a retrievable way, this is a wrapper around multiples ConcurrentHashMap
   # acting as a tree like structure.
   class MetricStore
-    class NamespacesExpectedError < Exception; end
+    class NamespacesExpectedError < StandardError; end
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: +1 on this.

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 12]
```diff
@@ -8,7 +8,14 @@ module LogStash module Instrument
...
+    class MetricNotFound < StandardError; end
+
+    KEY_PATH_SEPARATOR = "/"
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: can we freeze constants ? :+1:

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 13]
```diff
@@ -28,49 +35,107 @@ def fetch_or_store(namespaces, key, default_value = nil)
...
       end
+      return new_hash
     end
 
+    def extract_filter_keys(key)
+      key.to_s.strip.split(FILTER_KEYS_SEPARATOR).map(&:to_sym)
+    end
+
+    def transform_to_array(map)
+      map.values.collect do |value|
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: I wonder about the output here, are we really only interested on map values?

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 14]
```diff
@@ -28,49 +35,107 @@ def fetch_or_store(namespaces, key, default_value = nil)
...
+        value.is_a?(Hash) ? transform_to_array(value) : value
+      end.flatten
+    end
+
+    def transform_to_hash(map, new_hash = Hash.new({}))
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: I saw this notation ```Hash.new({})``` also in other places? any reason to use it? I mean you could just just ```Hash.new``` or ```{}```, right?

File: logstash-core/lib/logstash/instrument/metric_store.rb [hunk 15]
```diff
@@ -28,49 +35,107 @@ def fetch_or_store(namespaces, key, default_value = nil)
...
+      map.each_pair do |key, value|
+        if value.is_a?(Concurrent::Map)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: can we leverage the internal ```to_hash``` method provided by concurrent map?

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 14 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: not sure I like this name, also the ```transform_to_array```, are we transforming or cloning? we don't really do any data transformation, just data structure, right? also why not just using the typical ruby ```to_array``` or ```to_hash```?

File: logstash-core/lib/logstash/instrument/metric_type/base.rb [hunk 16]
```diff
@@ -1,24 +1,24 @@
...
-    def to_event(created_at = Time.now)
-      LogStash::Event.new(to_hash.merge({ "@timestamp" => created_at }))
+    def inspect
+      "#{self.class.name} - namespaces: #{namespaces} key: #{key} value: #{value}"
     end
 
-    def inspect
-      "#{self.class.name} - namespaces: #{@namespaces} key: #{@key} value: #{value}"
+    def to_json
+      LogStash::Json.dump(value)
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_type/base.rb]: +1

File: logstash-core/lib/logstash/instrument/metric_type/counter.rb [hunk 17]
```diff
@@ -25,14 +25,5 @@ def execute(action, value = 1)
     def value
       @counter.value
     end
-
-    def to_hash
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_type/counter.rb]: why not having this info in metric_type base? I see this as a common structure, at least for now for the types, right?

File: logstash-core/lib/logstash/instrument/namespaced_metric.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    # @param metric [LogStash::Instrument::Metric] The metric instance to proxy
+    # @param namespace [Array] The namespace to use
+    def initialize(metric, namespace_name)
+      @metric = metric
+      @namespace_name = Array(namespace_name)
+    end
+
+    # Get only the instance methods defined in the class
+    # and not the whole hierarchy of methods.
+    LogStash::Instrument::Metric.public_instance_methods(false).each do |method|
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/namespaced_metric.rb]: is there any specific reason not to use common inheritance patterns? why not using composition, so we could leverage  metric class methods without exposing the things we don't want ? what do you think?

File: logstash-core/lib/logstash/instrument/null_metric.rb [hunk 19]
```diff
@@ -2,41 +2,39 @@
...
-
-   # Allow to reuse the same variable when creating subnamespace
-   NULL_METRIC_INSTANCE = NullMetric.new
+    attr_reader :namespace_name
 
-   attr_reader :collector, :namespace_information
-   def initialize
+   # Get only the instance methods defined in the class
+   # and not the whole hierarchy of methods.
+   LogStash::Instrument::Metric.public_instance_methods(false).each do |method|
```
Comment (purbon) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: same thing here, why not using composition?

File: logstash-core/lib/logstash/pipeline.rb [hunk 20]
```diff
@@ -211,17 +212,22 @@ def start_workers
   end
 
   # Main body of what a worker thread does
-  # Repeatedly takes batches off the queu, filters, then outputs them
+  # Repeatedly takes batches off the queue, filters, then outputs them
   def worker_loop(batch_size, batch_delay)
     running = true
 
+    namespace_events = metric.namespace([:stats, :events])
```
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: If I understand properly, stats namespace gets initialized here, right? I guess is better to have it at metrics creation time, isn't? what do you think?

File: logstash-core/lib/logstash/pipeline.rb [hunk 21]
```diff
@@ -412,12 +423,14 @@ def shutdown_workers
   def plugin(plugin_type, name, *args)
     args << {} if args.empty?
 
+    pipeline_scoped_metric = metric.namespace([:stats, :pipelines, pipeline_id.to_s.to_sym, :plugins])
```
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: why the ```to_s.to_sym``` chained conversion? any specific fear to chain them?

File: logstash-core/lib/logstash/plugin.rb [hunk 22]
```diff
@@ -92,7 +92,7 @@ def debug_info
   end
 
   def metric=(new_metric)
-    @metric = new_metric.namespace(@id)
+    @metric = new_metric
```
Comment (purbon) [inline logstash-core/lib/logstash/plugin.rb]: This is interesting change, you moved the responsibility to namespace the metric with plugin id from the same plugin to the user of this plugin interface, isn't ? any reason for this? I see this as a more natural behaviour of the plugin base class as we aim to have all plugins request namespaced.

**Comment by purbon** [comment]
@ph did a first round of reviews, good work man! happy to chat with you more in deep if necessary to explain context of your changes :+1:

File: logstash-core/lib/logstash/agent.rb (hunk 1 above)
Comment (ph) [inline logstash-core/lib/logstash/agent.rb]: Gif diff are weird sometime, you know you can pass `?w=1` to the PR url,  this will make the diff ignore all the whitespace changes. The downside you cannot comment on the pr in that context... We can't have all the nice things.

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (ph) [inline logstash-core/lib/logstash/agent.rb]: This is effectively a good catch I will recheck that.

File: logstash-core/lib/logstash/filter_delegator.rb (hunk 3 above)
Comment (ph) [inline logstash-core/lib/logstash/filter_delegator.rb]: It can happen in any filters that does some kind of internal buffering like the multiline filter.
I can make the comment more generic if its the issues? I wanted to add context for the reviewer for why I was doing that check in this case.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 4 above)
Comment (ph) [inline logstash-core/lib/logstash/inputs/metrics.rb]: Yes we arent, but not having this make logstash display a yellow warning on boot, I have another PR #4603 that take care of that problem.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 5 above)
Comment (ph) [inline logstash-core/lib/logstash/inputs/metrics.rb]: It could but If I remember correctly LogStash::Json.dump(hash) will call to_hash on the object before calling `to_json`, I will double check it.

File: logstash-core/lib/logstash/instrument/collector.rb (hunk 6 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/collector.rb]: It should be in fact exposed to the user when the configuration is done, so right now I think using 1 is not a real issue until I have benchmarked the code.

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 7 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: Yes, but this is intended for custom internal pipeline. Lets say you want to update 2 different keys in the same code context.

```
stats.pipelines.events_in
stats.pipelines.pipelines01.plugins.XXXXX.events_in
```

This require 2 different namespace, I think is an OK trade off, we protect the plugins author to create metrics anywhere they want, but we are still able to do it.

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 8 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: Yes!

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: This is related to the with change come great responsabilities and the fact that we can create different keys in different context in the pipeline.

This is just a simple factory to use the collector singleton.

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 10 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: Private here actually mean this: I don't want the external world to be able to create this class.
But this doesn't prevent the class from returning an instance of it which we can interact with.

We have a test covering that specific use case if you look at the rspec.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 12 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: We we can, I always forget to do it.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: We don't need it, since the metric itself knows what namespace and keys its associated.
This make debugging a metric easier and more contained, the structure isn't really important.

I think using `#each` or `#get` depends on what you actually need to output to the user.
In the context on of the api we have a context of hierachy and how an metric rely to other metrics. but in a context of sending event to other system we only need to make sure that a metric has enough information to be meaningful to the external system.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 14 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: There is a small difference between `Hash.new()` and `Hash.new({})`, the argument is actually the default values for the keys, this make the code simpler because I don't have to worry about a default values for a key.

```ruby
jruby-1.7.23 :001 > m = Hash.new
 => {}
jruby-1.7.23 :002 > m[:j][:k]
NoMethodError: undefined method `[]' for nil:NilClass

jruby-1.7.23 :003 > m = Hash.new({})
 => {}
jruby-1.7.23 :004 > m[:j][:k] = "dd"

```

Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: I am not a big fan of `transform_*` name in that context , but for me `to_array` or `to_hash` actually mean I am converting the current object to a hash or an array which isn't exactly the case since it doesn't act on the object itself but on a generated subset.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 15 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Yes and no, it doesn't work with nested `Concurrent::Map`, It will just transform the first level.

File: logstash-core/lib/logstash/instrument/metric_type/counter.rb (hunk 17 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_type/counter.rb]: I can revisit this, but IIRC I had issue when mixing it with a `to_json` method.

I think `LogStash::Json.dump`  was calling to_hash before calling the `to_json` :(

File: logstash-core/lib/logstash/instrument/namespaced_metric.rb (hunk 18 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/namespaced_metric.rb]: Well, the signature of the method change, I can't use it :(

File: logstash-core/lib/logstash/instrument/null_metric.rb (hunk 19 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: This reduce the number of lines to code without affecting performance. Actually all theses methods are noop and need to exist.

Comment (purbon) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: and in the other side it makes readability much harder, I wonder if this method is actually faster or slower than pure composite here. What do you think?

File: logstash-core/lib/logstash/pipeline.rb (hunk 21 above)
Comment (ph) [inline logstash-core/lib/logstash/pipeline.rb]: If its a number, `to_sym` wont work, it only work on string. :(

File: logstash-core/lib/logstash/plugin.rb (hunk 22 above)
Comment (ph) [inline logstash-core/lib/logstash/plugin.rb]: Thanks for noticing it.

I have struggled with this change a bit, when I changed for a `NamespacedMetric` I though it was more natural to provide a context to the plugin. Since the plugins are actually instantiated in the delegators and can be run in multiples thread I though It was better generate the metric context there to make it more granular.

As a plugin developer he doesnt need to know what logic need to be done to have his keys to be correctly routed to the backend. 

Do you see any advantages to make it there? I am happy to discuss that!

File: logstash-core/lib/logstash/instrument/null_metric.rb (hunk 19 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/null_metric.rb]: - Agree on readability.
- I do not believe its slower, we actually use this pattern with the config ast and the filter function.
- Theses methods are created once, when the file is required.

File: logstash-core/lib/logstash/pipeline.rb (hunk 20 above)
Comment (ph) [inline logstash-core/lib/logstash/pipeline.rb]: Actually it will initialized at metrics creation time, its a lazy call.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 14 above)
Comment (guyboertje) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: @ph and @purbon - I have been bitten before on this form of Hash default - the same instance will be used for all missing keys. Make sure this does not cause problems...
```sh
irb(main):001:0> d = Hash.new
=> {}
irb(main):002:0> h = Hash.new(d)
=> {}
irb(main):003:0> h[:a][:b] = 42
=> 42
irb(main):004:0> h[:b][:c] = 43
=> 43
irb(main):005:0> d
=> {:b=>42, :c=>43}
irb(main):006:0> h[:b][:c]
=> 43
irb(main):007:0> h[:a][:c]
=> 43
irb(main):008:0> h[:z][:b]
=> 42
```

Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Good points!  Thanks @guy I will check that to make sure this is not a problem.

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (ph) [inline logstash-core/lib/logstash/agent.rb]: I rechecked the logic the namespace are created just outside the worker loop in the pipeline.
I understand the logic is a bit more harder to follow.

https://github.com/elastic/logstash/pull/4602/files#diff-bc1e9ea81caa54ea0d56abf4415fdf00R220

The problem is this part of the code actually update two different namespace and increment a counter. The pipeline specific and the application level counter for the number of `events_in`.

This is actually a local aggregation, to reduce complexity in the code base I preferred to update two different counter. What we could do in a future version is to allow to register aggregation callback or create aggregated view on metric store. I think this is a bit more complex to implement and it could be done in a future iteration.

File: logstash-core/lib/logstash/filter_delegator.rb (hunk 3 above)
Comment (ph) [inline logstash-core/lib/logstash/filter_delegator.rb]: Added a clearer comment on that part of the code.

File: logstash-core/lib/logstash/instrument/collector.rb (hunk 6 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/collector.rb]: Lets make it configurable in a new PR.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 5 above)
Comment (ph) [inline logstash-core/lib/logstash/inputs/metrics.rb]: Oh this is subtile, I found what problem I had before, Jrjackson does the serializing a bit different than the default `json` library, I think this is for performance reason @guyboertje ?

```ruby
require "json"
require "jrjackson"

class MyTest
  attr_reader :message

  def initialize(message)
    @message = message
  end

  def to_hash
    { "message" => "bye world" }
  end

  def to_json(opts = {})
    message
  end

  def to_json_data
    "to_json_data:  #{message}"
  end
end

class MyTest2
  attr_reader :message

  def initialize(message)
    @message = message
  end

  def to_hash
    { "message" => "bye world" }
  end

  def to_json(opts = {})
    message
  end
end

m = MyTest.new("Hello PH")
m2 = MyTest2.new("Hello PH 2")

puts JSON.dump(m)
puts JrJackson::Json.dump(m)
puts JrJackson::Base.generate(m, {})
puts JrJackson::Base.generate(m2, {})
```

```
Hello PH
Hello PH
"to_json_data:  Hello PH"
{"message":"bye world"}
```

If I override `to_json_data` instead of `to_json` it wont call `to_hash`, @guyboertje is there any specific reason why `JrJackson::Json.dump` and `JrJackson::Base.generate` dont call the same methods?

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 14 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: TIL, there was actually no error in the code, since in some case we were assigning an empty hash anyway, I have removed the default hash value.

File: logstash-core/lib/logstash/instrument/metric_type/counter.rb (hunk 17 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_type/counter.rb]: I move it back to the base class and used `to_json_data` in the base class.

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 5 above)
Comment (guyboertje) [inline logstash-core/lib/logstash/inputs/metrics.rb]: I will reply in Slack because this conversation is hidden under an outdated diff.

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (purbon) [inline logstash-core/lib/logstash/agent.rb]: If I understood your comment properly, you say that https://github.com/elastic/logstash/pull/4602/files#diff-bc1e9ea81caa54ea0d56abf4415fdf00R220 is basically creating the namespace?

File: logstash-core/lib/logstash/inputs/metrics.rb (hunk 4 above)
Comment (purbon) [inline logstash-core/lib/logstash/inputs/metrics.rb]: good for #4603 !!

File: logstash-core/lib/logstash/instrument/collector.rb (hunk 6 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/collector.rb]: +1

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 7 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: Ok! I understand. Good we are not passing this to plugins, can we might be add a validation/defense to check the namespace actually exist? I see a potential error where people actually create metrics to unknown locations.

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: I understand, as I shared earlier I think I would add some kind of defense/validation to prevent people falling on creating unexpected namespaces, what do you think?

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 10 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: :smile:

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Interesting,  my question here goes more on the side of if I ask for namespace `foo.bar` i might expect to also get data for `foo.bar.zoo`, what do you think?

File: logstash-core/lib/logstash/instrument/metric_type/counter.rb (hunk 17 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_type/counter.rb]: @ph might be need to fix `LogStash::Json.dump` any reason to do that like this?

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: Can we do it as the part of the #4506 ?

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: lets say that you have the following structure
```
{ :foo => { :bar = > { :zoo => 1, :hello => { :ninja => 2 } } }
```
If you do `each("foo/bar")` you will get [1, 2] since they are both under `foo.bar`

File: logstash-core/lib/logstash/pipeline.rb (hunk 20 above)
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: so you mean this will create the namespace first time we actually store something? what happen then for things with value zero?

File: logstash-core/lib/logstash/pipeline.rb (hunk 21 above)
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: makes sense

File: logstash-core/lib/logstash/pipeline.rb (hunk 20 above)
Comment (ph) [inline logstash-core/lib/logstash/pipeline.rb]: We dont have values until its gathered?

File: logstash-core/lib/logstash/agent.rb (hunk 2 above)
Comment (ph) [inline logstash-core/lib/logstash/agent.rb]: The namespace will only be created if we actually collect a metric `increment` in that case.

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 7 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: I suggest we make it part of #4506

File: logstash-core/lib/logstash/plugin.rb (hunk 22 above)
Comment (purbon) [inline logstash-core/lib/logstash/plugin.rb]: This is actually good question, my expectations somehow match with yours, plugins should not be aware of namespaces, they should just get the right metrics object. But if you ask me about internal logstash location for this change I think it belongs to the plugin base class. Having this change in the delegators makes you have duplicated code and introduced potential issues with plugins when the metric object is passed without the "right" namespace. What do you think?

File: logstash-core/lib/logstash/pipeline.rb (hunk 20 above)
Comment (purbon) [inline logstash-core/lib/logstash/pipeline.rb]: I know that, just wondering what will be the expected output and behaviour for a metric namespace created, but without any data. 

I know now we create the namespace the same time we insert the data, so having no values stored means no namespace created. Do you think it could be the case we aim to show a metric with value zero?

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 7 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: +1

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric.rb]: sure, just update the issue with it.

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: this makes this confusing, isn't?

**Comment by purbon** [comment]
I'm good to merge this, if we open two issues to fix in next PR's this features:

* https://github.com/elastic/logstash/pull/4602#discussion_r51553770
* https://github.com/elastic/logstash/pull/4602#discussion_r51736794
* https://github.com/elastic/logstash/pull/4602#discussion_r51736491

what do you think?

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: Well, It depends, it you need to retain structure you can use `#get` instead of each?

File: logstash-core/lib/logstash/instrument/metric.rb (hunk 9 above)
Comment (ph) [inline logstash-core/lib/logstash/instrument/metric.rb]: +1 awesome

File: logstash-core/lib/logstash/instrument/metric_store.rb (hunk 13 above)
Comment (purbon) [inline logstash-core/lib/logstash/instrument/metric_store.rb]: I see both providing equivalent features, if the two of them have different output we should be sure to document this very clear, otherwise make the api confusing.
